### PR TITLE
Support alternate runtime for host privileged operations

### PR DESCRIPTION
**Pull request by sameo**
Some CRI-O runtimes (like [cc-oci-runtime](https://github.com/01org/cc-oci-runtime)) do not support many host privileged operations like e.g. giving access to the host namespaces or running fully privileged containers (with access to *all* host devices). Those runtimes usually provide a higher level of container security by e.g. running container workloads within virtual machines and therefore running host privileged containers with them makes little sense.

This pull request tries to overcome that problem by allowing `ocid` users to define a *host privileged capable runtime path* in addition to the default runtime path. When `ocid` gets a request for creating a container or a sandbox with either the privileged flag set or access to at least one of the host namespaces (PID, IPC or networking), it will check if a host privileged runtime is defined and use it if it is. It will obviously use the default runtime if the host privileged one is not defined.

This PR only checks for the pod security context and each container within a given pod will inherit its pod privileged flags. In other words, that means we will run the privileged runtime in either one of those 2 cases:

1. The sandbox supports running at least one privileged container.
2. The sandbox requires access to either one of the host PID, IPC or networking namespace.

##### Note
I realize this **might** be fixed with [CRI multiple runtimes](https://github.com/kubernetes/kubernetes/pull/40662) support. If/when this happens, we should be able to remove part of this PR code.

cc @feiskyer @mcastelino

### CODE REVIEW

File: server/container_create.go [hunk 1]
```diff
@@ -286,9 +313,18 @@ func (s *Server) createSandboxContainer(ctx context.Context, containerID string,
...
+		rspec.MountNamespace:  "mount",
+		rspec.IPCNamespace:    "ipc",
+		rspec.UTSNamespace:    "uts",
+		rspec.UserNamespace:   "user",
+		rspec.CgroupNamespace: "cgroup",
+	} {
+		nsPath := fmt.Sprintf("/proc/%d/ns/%s", podInfraState.Pid, nsFile)
+		if err := specgen.AddOrReplaceLinuxNamespace((string)(nsType), nsPath); err != nil {
+			return nil, err
+		}
```
Comment (cyphar) [inline server/container_create.go]: I don't think this is correct. Different containers in a Pod do **not** share `mount`, `pid` or `cgroup` namespaces. And I'm fairly sure they don't share `uts` or `user` namespaces either.

What is the reason for this change?

**Comment by cyphar** [comment]
I'm a bit confused why this logic has to live in `cri-o`. Is there a reason why alternative runtimes can't do one of the following:

1. Look at the `config.json` that they've been given, and then figure out whether they have to run in "privileged" or "unprivileged" mode (effectively a wrapper).

2. Implement this "is this operation privileged" logic in `cri-o`, but expose it as an annotation rather than including two different runtime paths inside `cri-o`.

File: server/container_create.go (hunk 1 above)
Comment (jawnsy) [inline server/container_create.go]: That nsPath looks like it should use filepath.Join instead? Also `string(nsType)` instead of `(string)(nsType)`? (I didn't even realize the latter is even valid Go code?)

Comment (cyphar) [inline server/container_create.go]: `(string)` -> `string` is why it works.

But that's not the main point I'm making. I'm saying that the logic behind the change doesn't make sense to me (different containers in a sandbox do **not** share all of their namespaces).

Comment (mikebrow) [inline server/container_create.go]: Looks like the direction is to by default have containers in a pod share network, PID and IPC namespaces.

https://github.com/kubernetes/community/pull/207

Caught me by surprise the other day as well.  But it's switchable.

**Comment by sameo** [comment]
@cyphar 

> But that's not the main point I'm making. I'm saying that the logic behind the change doesn't make sense to me (different containers in a sandbox do not share all of their namespaces).

You're correct, the default k8s behavior is to share the PID, IPC and networking namespace, I'll fix that patch.

**Comment by sameo** [comment]
@cyphar

> I'm a bit confused why this logic has to live in cri-o. Is there a reason why alternative runtimes can't do one of the following:
>
> 1.  Look at the config.json that they've been given, and then figure out whether they have to run in "privileged" or "unprivileged" mode (effectively a wrapper).
>
> 2.    Implement this "is this operation privileged" logic in cri-o, but expose it as an annotation rather than including two different runtime paths inside cri-o.

Good points. Solution 1 is something I thought about. However, running privileged containers with any hypervisor based runtime is not possible (You can't see or change the host network namespace for example) so this would basically mean calling another runtime (e.g. `runc`) from those runtimes.
This seems a bit wrong to me and to some extend the same logic could be applied to `runc` for supporting multiple runtimes: Depending on the potential annotations it'd get from kubelet/CRI it would fork different runtimes when it does not support the incoming annotation (In case we get a declarative, intent annotation like e.g. `kubernetes/untrusted_workload=true`).

**Comment by feiskyer** [comment]
+1 for the direction. 

> Due to CRI not clearly defining the pod and container security context ordering (Right now, the container security context can be more permissive than the pod one) this PR checks for both the pod and container security context although ideally it should only check for the pod one. I'm filling a k8s issue for this.

I think we should only switch to privileged runtime in these two conditions:

* privileged sandbox and its containers (LinuxSandboxSecurityContext.Priviledged = true)
* HostNetwork, HostIPC and HostPid sandbox and its containers (LinuxSandboxSecurityContext. NamespaceOption is set)

**Comment by sameo** [comment]
@feiskyer 

> I think we should only switch to privileged runtime in these two conditions:
>  *  privileged sandbox and its containers (LinuxSandboxSecurityContext.Priviledged = true)
>  *  HostNetwork, HostIPC and HostPid sandbox and its containers (LinuxSandboxSecurityContext. NamespaceOption is set)

Ok, then we assume that the container security context will always be the same as the pod one. As you commented on [k8s issue 41848](https://github.com/kubernetes/kubernetes/issues/41848), kuberuntime currently sets all containers security contexts to be the same as their pod.

Right now this PR checks for both the containers and the pod security contexts, I'll fix the code to follow the assumption that they will always be identical.

**Comment by feiskyer** [comment]
> commented on k8s issue 41848, kuberuntime currently sets all containers security contexts to be the same as their pod.

Clarify:  I didn't mean LinuxContainerSecurityContext is always same with LinuxSandboxSecurityContext, I mean their HostNetwork fields are always same. Other fields could still be different. (https://github.com/kubernetes/kubernetes/issues/41848#issuecomment-281626145). Sorry for the misleading.

**Comment by sameo** [comment]
@feiskyer 

> Ok, then we assume that the container security context will always be the same as the pod one. As you commented on k8s issue 41848, kuberuntime currently sets all containers security contexts to be the same as their pod.

After further private discussions, this is incorrect. The security context will not always be the same, only the `HostNetwork` field.

And as you described, we should run a privileged runtime for all containers if `sandbox.privileged` is set to true. In that case, the pod could contain a mix of privileged and unprivileged containers but we can't mix them together.

With that understanding, I'll fix my PR.

**Comment by sameo** [comment]
@feiskyer PR updated according to the latest comments.

**Comment by feiskyer** [comment]
@sameo LGTM.

cc/ @cyphar @mrunalp PTAL

File: server/server.go [hunk 2]
```diff
@@ -50,6 +52,64 @@ type Server struct {
 	appArmorProfile string
 }
 
+func ociPrivileged(spec *rspec.Spec) bool {
```
Comment (mrunalp) [inline server/server.go]: Can we add a comment here? Also, shouldn't we also check for uid = 0?

File: server/container_create.go [hunk 3]
```diff
@@ -286,9 +295,14 @@ func (s *Server) createSandboxContainer(ctx context.Context, containerID string,
 
 	logrus.Debugf("pod container state %+v", podInfraState)
 
-	ipcNsPath := fmt.Sprintf("/proc/%d/ns/ipc", podInfraState.Pid)
-	if err := specgen.AddOrReplaceLinuxNamespace("ipc", ipcNsPath); err != nil {
-		return nil, err
+	for nsType, nsFile := range map[rspec.NamespaceType]string{
+		rspec.PIDNamespace: "pid",
```
Comment (mrunalp) [inline server/container_create.go]: We can take this out for now as k8s is still discussing viability and it may be pushed to next release.

**Comment by mrunalp** [comment]
How do we handle the case where sandbox privileged is not triggered but a container is privileged? Isn't it too late?

**Comment by mcastelino** [comment]
@mrunalp @sameo it would be good if we could specify the high water mark for privileges which can be inferred from the POD definition. Having such a high water mark will prevent the POD from being in a partially created state.

File: server/server.go (hunk 2 above)
Comment (jawnsy) [inline server/server.go]: Isn't `--privileged` different from just running as uid 0?

**Comment by feiskyer** [comment]
> How do we handle the case where sandbox privileged is not triggered but a container is privileged? Isn't it too late?

@mrunalp Kubelet ensures sandbox privileged is set when at least one of containers belonging to it is privileged.

File: server/server.go (hunk 2 above)
Comment (sameo) [inline server/server.go]: @mrunalp I added some comments to the routine, hopefully they make sense.

File: oci/oci.go [hunk 4]
```diff
@@ -34,24 +34,26 @@ const (
 )
 
 // New creates a new Runtime with options provided
-func New(runtimePath string, conmonPath string, conmonEnv []string, cgroupManager string) (*Runtime, error) {
+func New(runtimePath string, runtimeHostPrivilegedPath string, conmonPath string, conmonEnv []string, cgroupManager string) (*Runtime, error) {
```
Comment (jawnsy) [inline oci/oci.go]: Design question: would it make sense to pass in a struct with configuration instead? Every time a new parameter is added (as today), the method and all its callers have to be adjusted, and it's annoying for optional parameters (because all the call sites will need to be changed to pass in zero values like "")

Have we considered passing in configuration as a special struct type, like:

```go
type RuntimeConfig struct {
  Name string
  Path string
  ...
}

func NewRuntime(config *RuntimeConfig) (*Runtime, error) {
  r := &Runtime{
    config: config,
  }
  return r, nil
}
  ...
}
```

(sidenote, what's the point in `error` as a second return type, if errors can never occur?)

**Comment by mrunalp** [comment]
> @mrunalp Kubelet ensures sandbox privileged is set when at least one of containers belonging to it is privileged.
@feiskyer @sameo Does this mean that we just need to check for sandbox privileged flag and there is no need to look at the container config?

**Comment by feiskyer** [comment]
> Does this mean that we just need to check for sandbox privileged flag and there is no need to look at the container config?

I think so. We should only check sandbox config to decide the privileged runtime.

**Comment by sameo** [comment]
@mrunalp Simplified the code through a pod annotation, as discussed during the ocid meeting.

**Comment by mrunalp** [comment]
LGTM

**Comment by feiskyer** [comment]
LGTM
